contrib/libxls/utils/utils_libxls.cc

```diff
@@ -0,0 +1,149 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "contrib/libxls/utils/utils_libxls.h"
+
+#include <fstream>
+#include <iostream>
+#include <string>
+
+#include "contrib/libxls/sandboxed.h"
+
+absl::Status GetError(LibxlsApi* api, xls_error_t error_code) {
+  SAPI_ASSIGN_OR_RETURN(const char* c_errstr, api->xls_getError(error_code));
+  sapi::v::RemotePtr sapi_errstr(const_cast<char*>(c_errstr));
+  SAPI_ASSIGN_OR_RETURN(std::string errstr,
+                        api->GetSandbox()->GetCString(sapi_errstr));
+
+  return absl::UnavailableError(errstr);
+}
+
+absl::StatusOr<LibXlsWorkbook> LibXlsWorkbook::Open(LibxlsSapiSandbox* sandbox,
+                                                    const std::string& filename,
+                                                    const std::string& encode) {
+  if (sandbox == nullptr) {
+    return absl::UnavailableError("Sandbox has to be defined");
+  }
+
+  LibxlsApi api(sandbox);
+
+  sapi::v::IntBase<xls_error_t> sapi_error;
+  sapi::v::CStr sapi_filename(filename.c_str());
+  sapi::v::CStr sapi_encode(encode.c_str());
+
+  SAPI_ASSIGN_OR_RETURN(
+      xlsWorkBook * wb,
+      api.xls_open_file(sapi_filename.PtrBefore(), sapi_encode.PtrBefore(),
+                        sapi_error.PtrAfter()));
+
+  if (wb == nullptr) {
+    return GetError(&api, sapi_error.GetValue());
+  }
+
+  sapi::v::Struct<xlsWorkBook> sapi_wb;
+  sapi_wb.SetRemote(wb);
+  SAPI_RETURN_IF_ERROR(sandbox->TransferFromSandboxee(&sapi_wb));
+  sapi_wb.SetRemote(nullptr);
+
+  return LibXlsWorkbook(sandbox, wb, sapi_wb.data().sheets.count);
+}
+
+void LibXlsWorkbook::Close() {
```

Comment:
Use RAII.
Should define dtor and move ctors.

Reply:
It's not a problem that we are returning object?
I believe the dtor is called and the memory in sandbox is freed.

Reply:
not if you define custom move ctor and move assignment (copying not supported)
```
LibXlsWorkbook(LibXlsWorkbook&& other) {
  *this = other;
}
LibXlsWorkbook& operator=(LibXlsWorkbook&& other) {
  using std::swap;
  if (this != &other) {
    swap(rsw_, other.rws_);
    swap(sandbox_, other.sandbox_);
    other.Close();
  }
  return *this;
}
```

Reply:
Then again this is not needed if you make `sapi::v::Struct` a member of `LibXlsWorkbook` (which will make it non-copyable and non-movable), yet it'll be still possible to return it as a `absl::StatusOr<LibXlsWorkboo>` due to mandatory RVO in C++17 (see https://en.cppreference.com/w/cpp/language/copy_elision)

Reply:
If I understand correctly the problem is that the `xlsWorkSheet` don't have a struct constructor, so we can't use it as member of class.
```/home/oshogbo/sapi-libxls/sandboxed_api/var_struct.h: In instantiation of 'sapi::v::Struct<T>::Struct(Args&& ...) [with Args = {sapi::v::Struct<xlsWorkSheet>}; T = xlsWorkSheet]':
/home/oshogbo/sapi-libxls/contrib/libxls/utils/utils_libxls.h:36:7:   required from here
/home/oshogbo/sapi-libxls/sandboxed_api/var_struct.h:32:72: error: no matching function for call to 'xlsWorkSheet::xlsWorkSheet(sapi::v::Struct<xlsWorkSheet>)'
   32 |   explicit Struct(Args&&... args) : struct_(std::forward<Args>(args)...) {
      |                                                                        ^
In file included from /home/oshogbo/sapi-libxls/contrib/libxls/sandboxed.h:23,
                 from ../test/libxls_test.cc:15:
./sapi_libxls.sapi.h:304:16: note: candidate: 'xlsWorkSheet::xlsWorkSheet()'
  304 | typedef struct xlsWorkSheet
      |                ^~~~~~~~~~~~
```

Reply:
you're probably trying to:
```
sapi::v::Struct<xlsWorkSheet> worksheet;
sapi::v::Struct<xlsWorkSheet> copy(worksheet); // error here
```
that won't work as `sapi::v::Struct` does not provide a copy ctor. It also does not have a move ctor.

What you can do is:
```
class LibXlsWorkbook {
public:
 static absl::StatusOr<LibXlsWorkbook> Open(const std::string& filename) {
   xlsWorkBook* rwb;
   // ...
   return absl::StatusOr<LibXlsWorkbook>(absl::in_place, rwb);
 }
private:
 LibXlsWorkbook(xlsWorkBook* rwb) {
   workbook_.SetRemote(rwb);
 }
 sapi::v::Struct<xlsWorkBook> workbook_;
}
```

Reply:
Nope this doesn't seems to work.
It's not a problem:
```
std::forward<Args>(args)... 
```
in sapi::v::struct.
Right now I simplified code to not do any copy (return) and the issue is with ctor of LibXlsWorkSheet